### ISSUE TITLE
plaidml-setup hangs on Clover (OpenCL 1.1)

### ISSUE THREAD

**Issue by Bengt** [plaidml/plaidml#48]
My Ubuntu 16.04 has Mesa, LLVM and more from the X-Swat PPA. `clinfo` reports 

    $ clinfo
    [...]
    clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  Success (1)
    Platform Name                                 Clover
    Device Name                                   AMD TAHITI (DRM 2.49.0 / 4.10.0-37-generic, LLVM 5.0.0)
    [...]
    Number of platforms                               1
    [...]
    Platform Version                                OpenCL 1.1 Mesa 17.2.3
    Device Version                                  OpenCL 1.1 Mesa 17.2.3
    Device OpenCL C Version                         OpenCL C 1.1

When I run `plaidml-setup`, I have to "Enable experimental device support" and then it just hangs. `radeontop` shows no activity on the GPU. I cannot stop `plaidml-setup` using <kbd>CTRL</kbd>+<kbd>C</kdb>.

Termination would have been nice and an error message telling me that OpenCL 1.1 is not supported would have been helpful.

**Comment by jbruestle** [comment]
Hmm, we should support OpenCL 1.1, and certainly shouldn't lock the process.  We haven't done any testing with Clover, clearly we need to do some.  That said, I thought that Clover was CPU only, and the it looks like the device is a GPU.  Could you provide some more detail about your system, specifically OS, hardware, and how you installed your OpenCL stack (version, etc)?

Also, when you say that plaidml-setup hang, is this immediately after selecting experimental support, or after you select your device?

**Comment by znmeb** [comment]
Clover is buggy beyond much hope of redemption. <https://bugs.freedesktop.org/show_bug.cgi?id=99553>. If `clpeak` doesn't run on Clover, odds are your application won't either. Assuming you have an AMD card, there are hacks to get just the OpenCL part of the AMD proprietary driver installed on Ubuntu 16.04, Fedora 25/26 and CentOS / RHEL 7. It's also in the Arch User Repository, which is what I'm on.

**Comment by choongng** [comment]
Thanks @znmeb. @Bengt, does CLPeak work on your machine? It appears that the problems in Clover are beyond a workaround on our side. We'll leave this open, without the common benchmarks working (CLPeak) it's unlikely there's anything we can do about it but we would welcome help.

**Comment by znmeb** [comment]
If you want to go with Arch, here's the relevant Wiki page. I can verify that PlaidML works with the AMD Bonaire using this setup:

https://wiki.archlinux.org/index.php/AMDGPU

**Comment by Bengt** [comment]
@jbruestle I am running an AMD Radeon HD 7970 (aka R9 280X) under Ubuntu 16.04.3 LTS without the hardware enablement stack. So besides LLVM 5 instead of LLVM 4 and some minor version updates, my system should be vanilla. [See the full CLInfo report on gist.github.com](https://gist.github.com/Bengt/51167f2f101999ed4917daa573e8db44).

`plaidml-setup` hangs for me after enabling telemetry reporting:

```
$ plaidml-setup
[...]
Enable telemetry reporting? (y,n)[y]:y

Almost done. Multiplying some matrices...
Tile code:
  function (B[X,Z], C[Z,Y]) -> (A) { A[x,y : X,Y] = +(B[x,z] * C[z,y]); }
[hangs]
```

CLPeak will not compile on my system in its current configuration, because I do not have the AMD APP SDK installed:

```
CMake Error at CMakeLists.txt:62 (message):
  Could not find OpenCL include/libs.  Set OPENCL_ROOT to your OpenCL SDK.
  Download AMD APP SDK
  http://developer.amd.com/tools-and-sdks/heterogeneous-computing/amd-accelerated-parallel-processing-app-sdk/
  for x86/x64 or pocl http://pocl.sourceforge.net/ for ARM systems
```

@znmeb Thanks for your comments. Interesting to see how many other applications have trouble running under clover.

I guess from the view of an application developer, it would be best to decide whether or not clover should be supported or not. Regular testing should then ensure functionality or there should be detection mechanism in place and the user should get an error message when running on an unsupported OpenCL implementation.

**Comment by jbruestle** [comment]
@Bengt I started looking into adding better error handling for whatever is wrong with Clover, but I ran into installation issues and ended up moving on to other concerns.  I think right now this one is a bit too low priority for us to get to it anytime soon.  I'm not sure exactly why it's hanging rather than erroring for you.  Since I don't like to leave tickets pending indefinitely, I'm going to close that is won't-fix for now.  Sorry.
